### PR TITLE
Contracts consistency check 

### PR DESCRIPTION
**Pull request by Dominik1999**
## Description

Fixes: #5152

Please, describe what this PR does in detail:
- If it's a bug fix, detail the root cause of the bug and how this PR fixes it.
- If it's a new feature, describe the feature this PR is introducing and design decisions.
- If it's a refactoring, describe why it is necessary. What are its pros and cons in respect to the previous code, and other possible design choices.

## PR review check list

Quality check list that cannot be automatically verified.

- Safety
    - [ ] The changes respect the necessary conditions for safety (https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints)
-  Code quality
    - [ ] Error conditions are handled
    - [ ] Exceptions are propagated to the correct parent greenlet
    - [ ] Exceptions are correctly classified as recoverable or unrecoverable
- Compatibility
    - [ ] State changes are forward compatible
    - [ ] Transport messages are backwards and forward compatible
- Commits
    - [ ] Have good messages
    - [ ] Squashed unecessary commits
- If it's a bug fix:
    - Regression test for the bug
        - [ ] Properly covers the bug
        - [ ] If an integration test is used, it could not be written as a unit test
- Documentation
    - [ ] A new CLI flag was introduced, is there documentation that explains usage?
- Specs
    - [ ] If this is a protocol change, are the specs updated accordingly? If so, please link PR from the specs repo.
- Is it a user facing feature/bug fix?
    - [ ] Changelog entry has been added

### CODE REVIEW

**Comment by Dominik1999** [comment]
Might be WIP, we might add some tests and find a way to check if any contract is missing, right?

**Comment by Dominik1999** [comment]
> Might be WIP, we might add some tests and find a way to check if any contract is missing, right?

done, I moved a bit of code

File: raiden/ui/cli.py [hunk 1]
```diff
@@ -772,7 +772,7 @@ def print_step(description: str, error: bool = False) -> None:
             args["environment_type"] = environment_type
             args["extra_config"] = {"transport": {"matrix": {"available_servers": server_urls}}}
             args["one_to_n_contract_address"] = "0x" + "1" * 40
-            args["routing_mode"] = RoutingMode.PRIVATE
+            args["routing_mode"] = RoutingMode.LOCAL
```
Comment (rakanalh) [inline raiden/ui/cli.py]: I think you changed the default here to testing purposes, should it be reverted?

**Comment by hackaugusto** [comment]
Note about this PR:

This is how the smart contracts are related, as of 0.30.0 (dashed lines means the target smart contract is being deployed by the origin smart contract, so no validation is necessary here):

contracts

From the diagram above it is clear that one can use the `MonitoringService`, `OneToN`, or the `UserDeposit` as the entry point to the complete bundle of smart contracts. To simplify the command line this PR used the `UserDeposit`.

As one can see, there is a cycle among `UserDeposit`, `MonitoringService`, and `OneToN`. And multiple references to `Token`. Because of this many of these smart contracts have to be deploy and then initialized, which opens a window for mismatches in the addresses.

E.g. [The UserDeposit token](https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data_0.33.0/source/services/UserDeposit.sol#L56-L69), [the MonitoringService token](https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data_0.33.0/source/services/MonitoringService.sol#L60-L88), and [the ServiceRegistry token](https://github.com/raiden-network/raiden-contracts/blob/master/raiden_contracts/data_0.33.0/source/services/ServiceRegistry.sol#L232-L258), are all provided in the constructor as an argument, and can target different smart contracts. I have not thoroughly reviewed the code, it may be the case that this is not a problem, but it is a bit surprising, so these values also have to be checked.

To achieve the above this is what needs to be done:

- [x] Remove unnecessary command line arguments.
- [ ] Fetch the addresses of the other smart contracts through pointer chasing. (From the UD find the others).
- [ ] Validate the bytecode and internal consistency of the smart contract bundle.
  - To do the bytecode validation one has to add the missing proxies, since the bytecode is validated in the constructor of the proxies.
  - The best approach here is to have two new dataclasses, one for the `ServicesBundle` and another for the `RaidenNetwork`, the `__post_init__` of both these dataclasses should do the properties validation (e.g. checking all the properties match, like the token addresses mentioned above). If any of the proxies are not matching and exception should be raised and the object creation fail.

**Comment by Dominik1999** [comment]
also do that stuff with the Raiden services

**Comment by rakanalh** [comment]
Rebasing this PR

File: raiden/network/proxies/user_deposit.py [hunk 2]
```diff
@@ -92,6 +117,151 @@ def whole_balance_limit(self, block_identifier: BlockSpecification) -> TokenAmou
...
+        )
+        try:
+            existing_monitoring_service_address = self.monitoring_service_address(
+                block_identifier=given_block_identifier
+            )
+            existing_one_to_n_address = self.one_to_n_address(
+                block_identifier=given_block_identifier
+            )
+        except ValueError:
+            pass
```
Comment (karlb) [inline raiden/network/proxies/user_deposit.py]: I don't think we should silently ignore ValueErrors during the initialization. We should confirm that the contracts are initialized correctly before using them.

Comment (rakanalh) [reply raiden/network/proxies/user_deposit.py]: That's the pattern we use for pruned blocks as web3 would raise a `ValueError` when the block at `given_block_identifier` was pruned. In this case, we skip the precondition checks.

File: raiden/ui/app.py [hunk 3]
```diff
@@ -326,45 +348,37 @@ def run_app(
...
     message_handler = MessageHandler()
 
+    one_to_n_address = (
+        services_bundle.one_to_n.address if services_bundle.one_to_n is not None else None
+    )
+    monitoring_service_address = (
+        services_bundle.monitoring_service.address
+        if services_bundle.monitoring_service is not None
+        else None
+    )
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply raiden/ui/app.py]: if monitoring is disabled or routing is not set to `PFS` then we would not initialize the monitoring service nor OneToN... therefore the values would be None.

[the remark this answers is not in the document]
Comment (rakanalh) [reply raiden/ui/app.py]: See: https://github.com/raiden-network/raiden/pull/5430/files#diff-ff46da06c2592934425750ebadaec598R352

File: raiden/network/proxies/user_deposit.py (hunk 2 above)
Comment (karlb) [reply raiden/network/proxies/user_deposit.py]: I was in the wrong part of the code in my head. I thought we were skipping checks for correctly initialized contract addresses, which we should not do.

File: raiden/ui/app.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karlb) [reply raiden/ui/app.py]: I would prefer not to have all these `Optional`s even if it meant creating an unused proxy. But this is also nothing that belongs in this PR.

File: raiden/ui/startup.py [hunk 4]
```diff
@@ -33,6 +48,107 @@
...
+        if not token_address_matches_monitoring_service:
+            msg = (
+                f"The token used in the provided user deposit contract "
+                f"{user_deposit_address} does not match the one in the "
+                f"MonitoringService contract {monitoring_service_address}."
+            )
+            click.secho(msg, fg="red")
+            sys.exit(1)
+
+        token_address_matches_one_to_n = token_address == self.one_to_n.token_address(
```
Comment (czepluch) [inline raiden/ui/startup.py]: What is the point of this check? The `one_to_n` contract does not have a `token_address`. According to this PR there is a `token_address` function for the `one_to_n` proxy, but that function does not work.

**Pull request merged**